### PR TITLE
Implement didUpdateWidget for DefaultTabController

### LINKED ISSUE #18686: Hot reload doesn't recognize change in length in DefaultTabController
When updating the DefaultTabController from the [tutorial](https://flutter.io/cookbook/design/tabs/) to a length of 4 instead of 3 and reloading, Flutter doesn't recognize that the length value has changed. 
This is the code:

```dart
import 'package:flutter/material.dart';

void main() => runApp(new HeadacheApp());

class HeadacheApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: DefaultTabController(
        length: 4,
        child: Scaffold(
          appBar: AppBar(
            bottom: TabBar(
              tabs: [
                Tab(icon: Icon(Icons.directions_car)),
                Tab(icon: Icon(Icons.directions_transit)),
                Tab(icon: Icon(Icons.directions_bike)),
                Tab(icon: Icon(Icons.help)),
              ],
            ),
            title: Text('Tabs Demo'),
          ),
          body: TabBarView(
            children: [
              Icon(Icons.directions_car),
              Icon(Icons.directions_transit),
              Icon(Icons.directions_bike),
              Icon(Icons.help),
            ],
          ),
        ),
      ),
    );
  }
}
```

It gives the following error:

```
flutter: ══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
flutter: The following assertion was thrown while handling a gesture:
flutter: 'package:flutter/src/material/tab_controller.dart': Failed assertion: line 112 pos 12: 'value >= 0
flutter: && (value < length || length == 0)': is not true.
flutter:
flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
flutter: more information in this error message to help you determine and fix the underlying cause.
flutter: In either case, please report this assertion by filing a bug on GitHub:
flutter:   https://github.com/flutter/flutter/issues/new
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #2      TabController._changeIndex (package:flutter/src/material/tab_controller.dart)
flutter: #3      TabController.animateTo (package:flutter/src/material/tab_controller.dart:164:5)
flutter: #4      _TabBarState._handleTap (package:flutter/src/material/tabs.dart:864:17)
flutter: #5      _TabBarState.build.<anonymous closure> (package:flutter/src/material/tabs.dart:939:21)
flutter: #6      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:494:14)
flutter: #7      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:549:30)
flutter: #8      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:102:24)
flutter: #9      TapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:161:9)
flutter: #10     TapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:94:7)
flutter: #11     PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:315:9)
flutter: #12     PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:73:12)
flutter: #13     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:101:11)
flutter: #14     _WidgetsFlutterBinding&BindingBase&GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:143:19)
flutter: #15     _WidgetsFlutterBinding&BindingBase&GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:121:22)
flutter: #16     _WidgetsFlutterBinding&BindingBase&GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:101:7)
flutter: #17     _WidgetsFlutterBinding&BindingBase&GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:64:7)
flutter: #18     _WidgetsFlutterBinding&BindingBase&GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:48:7)
flutter: #19     _invoke1 (dart:ui/hooks.dart:134:13)
flutter: #20     _dispatchPointerDataPacket (dart:ui/hooks.dart:91:5)
flutter: (elided 2 frames from class _AssertionError)
flutter:
flutter: Handler: onTap
flutter: Recognizer:
flutter:   TapGestureRecognizer#8e887(debugOwner: GestureDetector, state: possible, won arena, finalPosition:
flutter:   Offset(329.3, 132.7), sent tap down)
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
```

I went into `flutter/src/material/tab_controller.dart` and printed the values of `length` and `value` which were both 3 instead of 3 and 4, which caused the exception to be raised. I was able to fix this by restarting the iPhone simulator and re-uploading the app.

### LINKED ISSUE #20292: DefaultTabController not working with dynamic data
## Steps to Reproduce

1. Add `DefaultTabController` widget 
2. add `length` to a dynamic property lets say `tabCount` from view model (_vm)
3. Load data asynchronously and based on that data generate `TabBarView` and `TabNames`
4. Click on tabs ( Error is thrown) also default active tab is set to index 1
5. If `tabCount` is set to a fixed value and `tabNames` are static values lets say `4` or `3` and `["A", "B"]` it works

## Code

  ```dart
@override
  Widget build(BuildContext context) {
    return new StoreConnector<AppState, _ViewModel>(
      onInit: (Store<AppState> store) => store.dispatch(initSportData(context)),
      converter: _ViewModel.fromStore,
      builder: (BuildContext context, _ViewModel vm) {
        return new DefaultTabController(
            length: vm.tabCount,  // <-------------------------------------------Here
            child: Scaffold(
              appBar: PreferredSize(
                  preferredSize: Size.fromHeight(96.0),
                  child: getAppBar(context, vm.tabNames)), // <--------------------------------------Here
              body: new Container(
                  decoration: new BoxDecoration(
                      image: new DecorationImage(
                    alignment: Alignment.topLeft,
                    image: new AssetImage('images/main-bg.png'),
                    fit: BoxFit.cover,
                  )),
                  child: vm.isLoading
                      ? new Center(
                          child: new CircularProgressIndicator(),
                        )
                      : TabBarView(
                          children: getTabBarView(vm.tabDetails), //<------------------------------Here
                        )),
            ));
      },
    );
  }

```
## Logs

```
 ══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
I/flutter ( 7806): The following assertion was thrown while handling a gesture:
I/flutter ( 7806): 'package:flutter/src/material/tab_controller.dart': Failed assertion: line 110 pos 12: 'value >= 0
I/flutter ( 7806): && (value < length || length == 0)': is not true.
I/flutter ( 7806): 
I/flutter ( 7806): Either the assertion indicates an error in the framework itself, or we should provide substantially
I/flutter ( 7806): more information in this error message to help you determine and fix the underlying cause.
I/flutter ( 7806): In either case, please report this assertion by filing a bug on GitHub:
I/flutter ( 7806):   https://github.com/flutter/flutter/issues/new
I/flutter ( 7806): 
I/flutter ( 7806): When the exception was thrown, this was the stack:
I/flutter ( 7806): #2      TabController._changeIndex (package:flutter/src/material/tab_controller.dart)
I/flutter ( 7806): #3      TabController.animateTo (package:flutter/src/material/tab_controller.dart:162:5)
I/flutter ( 7806): #4      _TabBarState._handleTap (package:flutter/src/material/tabs.dart:864:17)
I/flutter ( 7806): #5      _TabBarState.build.<anonymous closure> (package:flutter/src/material/tabs.dart:939:21)
I/flutter ( 7806): #6      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:494:14)
I/flutter ( 7806): #7      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:549:30)
I/flutter ( 7806): #8      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:102:24)
I/flutter ( 7806): #9      TapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:161:9)
I/flutter ( 7806): #10     TapGestureRecognizer.acceptGesture (package:flutter/src/gestures/tap.dart:123:7)
I/flutter ( 7806): #11     GestureArenaManager.sweep (package:flutter/src/gestures/arena.dart:156:27)
I/flutter ( 7806): #12     _WidgetsFlutterBinding&BindingBase&GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:147:20)
I/flutter ( 7806): #13     _WidgetsFlutterBinding&BindingBase&GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:121:22)
I/flutter ( 7806): #14     _WidgetsFlutterBinding&BindingBase&GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:101:7)
I/flutter ( 7806): #15     _WidgetsFlutterBinding&BindingBase&GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:64:7)
I/flutter ( 7806): #16     _WidgetsFlutterBinding&BindingBase&GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:48:7)
I/flutter ( 7806): #17     _invoke1 (dart:ui/hooks.dart:134:13)
I/flutter ( 7806): #18     _dispatchPointerDataPacket (dart:ui/hooks.dart:91:5)
I/flutter ( 7806): (elided 2 frames from class _AssertionError)
I/flutter ( 7806): 
I/flutter ( 7806): Handler: onTap
I/flutter ( 7806): Recognizer:
I/flutter ( 7806):   TapGestureRec
```

```
flutter doctor -v
[✓] Flutter (Channel beta, v0.5.1, on Mac OS X 10.13.6 17G65, locale en-IN)
    • Flutter version 0.5.1 at /private/var/www/flutter
    • Framework revision c7ea3ca377 (2 months ago), 2018-05-29 21:07:33 +0200
    • Engine revision 1ed25ca7b7
    • Dart version 2.0.0-dev.58.0.flutter-f981f09760

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/rajeev/Library/Android/sdk/
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = /Users/rajeev/Library/Android/sdk/
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[!] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.4.1, Build version 9F2000
    ✗ Brew not installed; use this to install tools for iOS device development.
      Download brew at https://brew.sh/.

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 27.0.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[!] IntelliJ IDEA Ultimate Edition (version 2018.1.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.io/intellij-setup/#installing-the-plugins

[!] VS Code (version 1.25.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected devices (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.0.0 (API 26) (emulator)

! Doctor found issues in 3 categories.
```

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes flutter#20292
Fixes flutter#18686

Currently trying to update the number of tabs without forcing the state object to be destroyed via a ValueKey (or similar) can cause index out of bounds or related errors when the child tabs try to render

### CODE REVIEW

File: packages/flutter/lib/src/material/tab_controller.dart [hunk 1]
```diff
@@ -299,6 +300,19 @@ class _DefaultTabControllerState extends State<DefaultTabController> with Single
     );
   }
 
+  @override
+  void didUpdateWidget(covariant DefaultTabController oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.length != oldWidget.length) {
+      final int oldIndex = _controller._index;
+      _controller = TabController(
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/tab_controller.dart]: Need to dispose the old controller first, but: what about clients that still hold a reference to the old controller?

Comment (jonahwilliams) [reply packages/flutter/lib/src/material/tab_controller.dart]: Opened https://github.com/flutter/flutter/pull/24783

File: packages/flutter/lib/src/material/tab_controller.dart [hunk 2]
```diff
@@ -299,6 +300,26 @@ class _DefaultTabControllerState extends State<DefaultTabController> with Single
...
+        vsync: this,
+        length: widget.length,
+        initialIndex: math.min(oldIndex, widget.length - 1),
+      );
+      // This has to be done in a microtask since child widgets will attempt
+      // to remove their listeners in the didUpdateWidget method. This will fail
+      // if the controller has already been disposed.
+      scheduleMicrotask(() {
+        oldController.dispose();
+      });
```
Comment (Hixie) [inline packages/flutter/lib/src/material/tab_controller.dart]: I feel like our API is bad if we've ended up here.

Maybe TabController should be more mutable, so we don't have to change the tab controller instance?

**Comment by jonahwilliams** [comment]
I'm not going to have time to get to this in the next few weeks, closing for now

**Pull request closed without merging**